### ISSUE TITLE
Missed optimization with movzx and mov

### ISSUE THREAD

**Issue by Hintro** [llvm/llvm-project#56498]
I came across this when examining a loop that runs slower than I expected. It involves explicit and implicit conversions between 8-bit and 32/64-bit values, and as I looked through the generated assembly using Godbolt compiler explorer, I found lots of movzx instructions that don't seem to break dependency or play a role in correctness, not to mention many use the same register like `movzx eax al`, which cannot be eliminated.

I then tried some simple examples on Godbolt, and found that this behavior is persistent and easily reproducible, even when I specify `-march=skylake`. Here's an example:
```cpp
#include <stdint.h>

int add2bytes(uint8_t* a, uint8_t* b) {
    return uint8_t(*a + *b);
}
```
Clang 14 `-O3`
```asm
add2bytes(unsigned char*, unsigned char*):                       # @add2bytes(unsigned char*, unsigned char*)
        mov     al, byte ptr [rsi]
        add     al, byte ptr [rdi]
        movzx   eax, al
        ret
```
`movzx` would be better in place of the `mov` instead of being at the end, so that dependency on old RAX value can be broken from the start and also clearing the upper bits of RAX in the process.

I also asked this on Stack Overflow and [Peter Cordes] has a detailed response (https://stackoverflow.com/a/72953035/14730360) explaining how this behavior is bad for pretty much all X86 processors. 

Godbolt link with code for examples: https://godbolt.org/z/z45xr4hq1
Here's one that's closer to what I was originally examining:
```cpp
int foo(uint8_t* a, uint8_t i, uint8_t j) {
    return a[a[i] | a[j]];
}
```
Clang 14 `-O3`:
```asm
foo(unsigned char*, unsigned char, unsigned char):                             # @foo(unsigned char*, unsigned char, unsigned char)
        mov     eax, esi
        mov     ecx, edx
        mov     cl, byte ptr [rdi + rcx]
        or      cl, byte ptr [rdi + rax]
        movzx   eax, cl
        movzx   eax, byte ptr [rdi + rax]
        ret
```
`movzx   eax, cl` here just seems unnecessary. The upper bits of RCX should already be clean as it is used as index in `mov     cl, byte ptr [rdi + rcx]`. The subsequent `or` does not affect its upper bits, and the dependency of RCX on this `or` is not something that `movzx eax cl` can break. So I think it's better to just do `movzx   eax, byte ptr [rdi + rcx]` after the `or`.

**Comment by topperc** [comment]
llvm turns 8-bit loads into movzx pretty late in `X86FixupBWInsts` pass. I think it only does if the code is in a loop because the movzx encoding is 1 byte longer.

**Comment by rotateright** [comment]
I drafted a patch that removes the loop restriction in X86FixupBWInsts, and there are 200+ regression test files that fail. So far, the diffs look as expected, but it's going to take some time to update everything and confirm.

**Comment by rotateright** [comment]
Proposal to use the zero-extending load outside of loops:
https://reviews.llvm.org/D129775

**Comment by RKSimon** [comment]
resolving - this was fixed in 15.x

**Comment by pcordes** [comment]
> resolving - this was fixed in 15.x

Actually only one part of this was fixed: using `movzx` for the initial write of a register.

We still get a (now useless) `movzx eax, al` at the end of `add2bytes` and `or2bytes` with Clang 20 and current trunk (https://godbolt.org/z/fPcf7PoxW).  Even with `-march=znver1` for example; AMD has never done any partial-register renaming.
   
Only P6-family Intel (PPro through Nehalem) should do that extra `movzx` to avoid a partial-register stall when the full register is read after an 8-bit write, but P6-fam is obsolete enough (~2010) that it should be fine to just never do it, even with `-mtune=nehalem`, if we don't want to extra code to handle both tuning choices.

Our `movzx eax, al` doesn't even avoid using the same register to allow mov-elimination (zero latency but still costs a uop for the front-end and takes a slot in the ROB (ReOrder Buffer = out-of-order exec window)).  `foo()` does, using `movzx eax, cl`.

Optimal asm for `or2bytes` is 
```
        movzx   eax, byte ptr [rsi]
        or      al, byte ptr [rdi]            # modify the low byte, leaving the upper 56 still zero
        ret
```

Or with `-march=nehalem` or other P6-family, `movzx`/`or` into ECX/CL and `movzx eax, cl`.  
(Actually on P6-family, no downside to only using AL / EAX, but a separate reg isn't worse on P6-family if there's one to spare.  And it is better for P6-tuned binaries running on newer CPUs.)

On P6-family, the cost of reading EAX after `or al, mem` is something like a 3 cycle stall to insert a merging uop on Core2/Nehalem, or like 6 cycles or more on earlier P6 that just stall and wait instead of inserting a merging uop, IIRC.  Agner Fog's microarch guide talks about this.  By stall I think he means stalling the front-end for that many cycles after issue/rename of an instruction that reads a split-renamed register.  
So it does suck quite badly for P6-family if we don't make `-march=nehalem` avoid this, but it's 15 years old now, and will be even older by the time a change makes it into a new Clang / LLVM version that gets into widespread use.

First-gen Sandybridge will insert a merging uop, so it basically does the `movzx` for us.  But actually `add/or al, mem` doesn't rename AL separately from RAX on SnB; those are 2-input ops that read their destination first, so partial-register renaming wouldn't avoid a dependency.  So the CPU doesn't do it for such instructions.  
There's also no cost on Haswell and later since AL can't get renamed separately from RAX in the first place on those uarches, only AH can get renamed separately.  
AMD CPUs never rename partial registers at all.

TL:DR: our current code-gen is wasting a `movzx eax, al` to avoid a performance pothole on Intel CPUs from 2010 and earlier (or because we just don't look for the optimization of narrow ops leaving zero-extended values zero-extended).  And in many cases we're not even doing it in the most efficient way possible (one that would allow mov-elimination).
